src/charm.py

```diff
@@ -47,6 +52,67 @@
 logger = logging.getLogger(__name__)
 
 
+def monotonize(value, default_value, multiplier):
```

Comment:
1. Is this method only used in the context of the Kubernetes Resource patch library ? If so can it be a utility method provided by the resource patch library ?
2. Are the function arguments expected to be non-negative ?
3. If arguments are non-negative, given what the code is doing would it make sense for the function signature to be ```
```
monotonize(default_value, multiplier, value = 0):
    assert default_vale >= 0
    assert multiplier >= 0
    assert multiplier < 1
    assert value >=0
```
5. The name of this function, its docstring and body do not seem to concur. The body suggests that this function is trying to do find something along the lines of a maximum (given above suggested signature) of "value", "default_value" and "multiplier*default_value". Is this correct ? Or alternatively could something like a maximum (or `find_maximum()`) be used instead of `monotonize()` ? Would this be more suggested of intended purpose ? The word "monotonize" suggests that this method takes in an unordered series and returns an ordered Monotonic series.
7. Is this really a [continuous function](https://en.wikipedia.org/wiki/Continuous_function) ? Or rather does the continuity of this function have any relevance to its purpose ? If not could the docstring  avoid any references to continuity ?

Reply:
1. If you think this is also going to be used in other places in COS then doesn't it make sense to provide it as a utility function in the Resource patch library.
2. The code below does not translate to the function definition of $y(x; y_0, 0 \lt k \lt 1)$ described above. For example the value `None` is not a mathematically meaningful value for $x$ or $y_0$ and $k$ which translates to `value` and `default_value` and `multiplier` in code where either can get a value `None`. Hence it is a bit misleading to think of body of `monotonize()` in terms of $y()$ since $x \leq \mathrm{None}$ or $y_0 \lt \mathrm{None} \leq y_0/k$ does not make mathematical sense. As shown in the code below `value` can be `None`. For these reasons I think the math here is a digression and we would be better of just looking at the code.
3. As I understand all of these "resource" and "limits" should never be negative because they represent real quantities that can not be negative.
4.  "monotonize" is bound to suggest to any cognizant person that this function transforms its arguments by putting them in a [monotonic ordering](https://en.wikipedia.org/wiki/Monotonic_function). But this function accepts multiple values and returns a single value (not a reordering of the same multiple input values). If you are indeed seeking a better name for this function then we could consider the purpose of this function in the context of the code were it is intended to be used rather than the loose mathematical analogy discussed above. As I understand this method is used in `limits_to_requests()` where it is used to constrain the Kubernetes resource "request" after taking into consideration the Kubernetes resource "limits" and a "scaling factor" (termed "multiplier"). If this understanding is correct then would a person reading this code find it easier to understand the intended purpose of this function if it were termed something like `constrain_k8s_request(limits, requests, scaling)` or something similar ?
5. This is probably a digression and not relevant to the intended purpose of this function and its code.

I do not mean to belabor these points, but I did struggle to understand the code. I will leave it to your better judgement.

Reply:
Moved to https://github.com/canonical/observability-libs/pull/19 and refactored a bit.
Renaming/removal of the function(s) there too is still pending.